daemon: move circular initialization of policy.Repository to hive

lmb opened the pull request:
IPCache, CachingIdentityAllocator and policy.Repository (really SelectorCache) have a gnarly circular dependency during intialization. Do the simplest possible thing and lift them into hive by providing a constructor that encapsulates the neccessary logic.

This way we can add dependencies on these without having to modularize them first.

Pre-requisite for #23046.

lmb committed:
egressgateway, ipcache: depend on k8s.CacheStatus

Introduce a small wrapper around chan struct{} to be able to
provide it via hive to egressgateway and ipcache.

Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

joamaki commented on daemon/cmd/policy.go:
```diff
@@ -58,20 +61,52 @@ func (d *Daemon) initPolicy(
+	secretManager certificatemanager.SecretManager,
+	nodeHandler ipcacheTypes.NodeHandler,
+	cacheStatus k8s.CacheStatus,
+) (CachingIdentityAllocator, *policy.Repository, *policy.Updater, *ipcache.IPCache, error) {
```
consider using ` cell.In` and `cell.Out` structs: https://docs.cilium.io/en/latest/contributing/development/hive/#provide.

joamaki commented on daemon/cmd/cells.go:
```diff
@@ -78,6 +80,8 @@ var (
 		// Auth is responsible for authenticating a request if required by a policy.
 		auth.Cell,
+
+		cell.Provide(newPolicyMess),
```
let's add a comment here to describe what this sets up (and maybe rename the function.. although I kind of like leaving it like this...)

joamaki commented on daemon/cmd/cells.go:
```diff
@@ -90,5 +94,9 @@ var (
 			newDatapath,
 		),
+
+		cell.Provide(func(dp datapath.Datapath) ipcacheTypes.NodeHandler {
```
I'd keep adaptions like this next to where they're used, e.g:

```
cell.Provide(
  func(dp datapath.Datapath) ipcacheTypes.NodeHandler { return dp.Node() },
  newPolicyMess,
)
```

Though here it would seem ok to me to just take `datapath.Datapath` in `newPolicyMess`.

joamaki reviewed (approved):
Yeah let's do this. Small incremental steps like this to unblock other work seems fine to me (rather than rabbit hole'ing into refactoring ipcache and policy repo :D)

lmb committed:
policy: add NewStoppedPolicyRepository

Add a constructor for Repository that doesn't allocate any queues.
This is necessary to satisfy hive invariants.

Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

gandro commented on daemon/cmd/policy.go:
```diff
@@ -58,20 +63,90 @@ func (d *Daemon) initPolicy(
+		OnStop: func(hc hive.HookContext) error {
+			cancel()
+			if err := ipc.Shutdown(); err != nil {
+				return err
```
Previously, we just logged this error and continued the shutdown (which basically seems only happens if ipc.Shutdown has been called already previously). Might be worth doing the same here.

lmb replied:
Good point, I misread it and thought that the log would panic. I've changed it so that the error is returned to hive instead, on the assumption that this will log something.

gandro commented on daemon/cmd/policy.go:
```diff
@@ -58,20 +63,90 @@ func (d *Daemon) initPolicy(
+			return nil
+		},
+		OnStop: func(hc hive.HookContext) error {
+			cancel()
```
While I don't think it actually matters in correctness, I believe it would be cleaner to call into the shutdown routines first, and then `cancel` the context. For example, some of the shutdown routines (like `ipc.Shutdown`) will wait for a controller to finish it's work first.

christarazi commented on pkg/policy/repository.go:
```diff
@@ -167,19 +167,27 @@ func NewPolicyRepository(
+//
+// Qeues must be allocated via [Repository.Start]. The function serves to
+// satisfy hive invariants.
+func NewStoppedPolicyRepository(
```
What does "Stopped" mean? I found out what it means by reading further along in the code, but it would be good to document it in this function's godoc.

joestringer replied:
```
$ git grep NewPolicyReposit | grep -v test.go
daemon/cmd/daemon.go:           // Must be done before calling policy.NewPolicyRepository() below.
pkg/policy/repository.go:// NewPolicyRepository creates a new policy repository.
pkg/policy/repository.go:func NewPolicyRepository(
$ git grep NewPolicyReposit | grep test.go | wc -l
68
```

It looks like the main NewPolicyRepository() is now just a helper for test code. It's a bit of a code smell to me to call this `NewStoppedPolicyRepository()` rather than just having a `NewPolicyRepository()` + `Start()`, but I appreciate that all those other test code references are a bit of a pain to fix up if they're all relying on this code as-is to initialize+start the policy repo. I'm OK to defer this for now.

lmb replied:
I've updated the function doc. I think Jussi's preference was to pass a `hive.Lifecycle` into the constructor, but that causes a big ripple effect, NewPolicyRepository is used a lot. So I did this botch to keep the changes small(er).

joestringer commented on daemon/cmd/policy.go:
```diff
@@ -58,20 +63,92 @@ func (d *Daemon) initPolicy(
+			idAlloc.Close()
+
+			// Cancel after shutting down IPCache.
+			cancel()
```
This was previously tied to the Daemon context and closed *before* `ipc.Shutdown()`, related to https://github.com/cilium/cilium/pull/21676. At the very least I think we should move this context up to close first -- basically this gives the early signal to start shutting down some controllers in `pkg/ipcache` which `Shutdown()` will want to wait on anyway.

joestringer replied:
(arguably we could maybe get rid of the `Context` here now if we're not trying to converge all the contexts onto the root tree of `Context`s any more, especially given that there's already an explicit `Shutdown()` call going on here, but that's more of an IPCache cleanup rather than what you're trying to achieve in this PR, so I'm happy to defer that work until later. For now I just would like to prevent unit test failures due to race conditions between goroutines shutting down)

lmb replied:
> I think we should move this context up to close first -- basically this gives the early signal to start shutting down some controllers in `pkg/ipcache` which `Shutdown()` will want to wait on anyway.

Ack, that's what I had in the beginning. Thanks for the context, I'll change it back.

joestringer replied:
(cc @gandro , I missed that you had conflicting suggestions but I do think that the cancel() needs to happen first here)

joestringer reviewed (changes requested):
Thanks, this looks like a nice step forward. Just one minor nit in something that I think functionally changed during the refactoring.

lmb committed:
daemon: move circular initialization of policy.Repository to hive

IPCache, CachingIdentityAllocator and policy.Repository (really
SelectorCache) have a gnarly circular dependency during initialization.
Do the simplest possible thing and lift them into hive by providing
a constructor that encapsulates the necessary logic.

This way we can add dependencies on these without having to modularize
them first.

During shutdown of the IPCache we first cancel the context
and then call IPCache.Shutdown, see cilium#21676.

Signed-off-by: Lorenz Bauer <lmb@isovalent.com>
